File: Interview Qs/Arrays/Median of Two Sorted Arrays/medianOfArrays.swift
// Ehud Adler
// 8.22.18
//
// Median
// https://leetcode.com/problems/median-of-two-sorted-arrays
// Runtime: 48ms
//

func findMedianSortedArrays(_ nums1: [Int], _ nums2: [Int]) -> Double {
  
    // Initial check. If we find 1 array empty the median
    // in the median of the other one
    if nums1.isEmpty {
        return nums2.findMedian()
    } else if nums2.isEmpty {
        return nums1.findMedian()
    }
    let merged = nums1.mergeSorted(with: nums2)
    return merged.findMedian()
}

extension Array where Element == Int {
  
  func mergeSorted(with arr: [Int]) -> [Int] {
    
    var runner1 = 0, runner2 = 0
    var merged = [Int]()
    while runner1 < self.count && runner2 < arr.count {
      merged.append(
        self[runner1] < arr[runner2]
          ? self[runner1.postIncrement()]
          : arr[runner2.postIncrement()]
      )
    }
    
    while runner1 < self.count {
      merged.append(self[runner1.postIncrement()])
    }
    
    while runner2 < arr.count {
      merged.append(arr[runner2.postIncrement()])
    }
    
    return merged
  }
  
  func findMedian() -> Double {
    let count = self.count, midPoint = self[(count / 2)]
    return count % 2 == 0
      ? Double((midPoint + self[(count / 2) - 1])) / 2.0
      : Double(midPoint)
  }
}

extension Int {
  mutating func postIncrement() -> Int {
    defer {self = self + 1}
    return self
  }
}

File: Interview Qs/Arrays/Median of Two Sorted Arrays/medianOfArrays.cpp
// Ehud Adler
// 8.23.18
//
// Median
// https://leetcode.com/problems/median-of-two-sorted-arrays
// Runtime: 16ms
//

class Solution {
    
vector<int> merge(const vector<int>&,const  vector<int>&);
double median(const vector<int>&);

public:
    double findMedianSortedArrays(
        vector<int>& nums1,
        vector<int>& nums2) {
            
        // Initial Check
        if (nums1.empty()) {
               return get_median(nums2);
        } else if (nums2.empty()) {
               return get_median(nums1);
        }
        vector<int> merged = merge_sorted(nums1, nums2);
        return get_median(merged);
    }

    vector<int> merge_sorted(
        const vector<int>& arr1,
        const vector<int>& arr2) {
        
        unsigned int runner1, runner2;
        runner1 = runner2 = 0;
        vector<int> merged;

        while (runner1 < arr1.size() && runner2 < arr2.size()) {
            merged.push_back(
                arr1[runner1] < arr2[runner2] 
                ? arr1[runner1++] 
                : arr2[runner2++]
            );
        }

        while (runner1 < arr1.size()) {
            merged.push_back(arr1[runner1++]);
        }

        while (runner2 < arr2.size()) {
            merged.push_back(arr2[runner2++]);
        }
        return merged;
    }

    double get_median(const vector<int>& arr) {
        unsigned int count = arr.size(),
        midPoint = count / 2;

        return count % 2 == 0 
        ? (arr[midPoint] +  arr[midPoint-1]) / 2.0
        : (double)arr[midPoint];
    }
};
